src/reducers/notifications.reducer.js

```diff
@@ -0,0 +1,39 @@
+import {
+  ADD_NOTIFICATION,
+  UPDATE_NOTIFICATION,
+  DELETE_NOTIFICATION,
+} from '../actions';
+
+import type { Action } from 'redux';
+import type { Notification } from '../types';
+
+type State = {
+  [uniqueNotificationId: string]: Notification,
+};
+
+const initialState = {};
+
+export default (state: State = initialState, action: Action) => {
+  switch (action.type) {
+    case ADD_NOTIFICATION:
+    case UPDATE_NOTIFICATION:
+      const existing = state[action.notificationId] || {};
+      return {
+        ...state,
+        [action.notificationId]: {
+          title: action.title || existing.title,
+          message: action.message || existing.message,
+          progress: action.progress || existing.progress,
+          complete: action.complete || existing.complete,
+        },
+      };
```

Comment:
I think these two cases should be handled separately :) even if it's a bit redundant.

Also, I like using Immer in reducers. It's a library that lets you pretend the state is mutable (and it uses proxies behind-the-scenes to do structural sharing, so it's possibly more performant... not that it matters, in this case)

```js
case ADD_NOTIFICATION:
  const { type, notificationId, ...notificationData } = action;

  return produce(state, draftState => {
    draftState[notificationId] = notificationData;
  })

case UPDATE_NOTIFICATION:
  const { type, notificationId, ...notificationData } = action;

  return produce(state, draftState => {
    const existing = draftState[notificationId];
    draftState[notificationId] = { ...existing, ...notificationData };
  })

```

I haven't run this, so not sure if it'd work, but yeah something like this matches our conventions more closely. In general I try and avoid combining branches unless they're identical